logging/src/logging_v2.cpp

```
@@ -95,283 +93,344 @@ bool emit_ack(std::shared_ptr<logging::metafilter_t> filter,
+            default_mf->add_filter(logging::filter_info_t(filter_info_conf));
+        }
+
+        auto core_mf = get_core_metafilter();
+        cocaine::filter_t core_filter([=](blackhole::severity_t sev, blackhole::attribute_pack& pack) {
+            logging::filter_result_t result;
+            if(core_mf->empty()) {
+                result = default_mf->apply(sev, pack);
+            } else {
+                result = core_mf->apply(sev, pack);
+            }
+            if(result == logging::filter_result_t::accept) {
+                return true;
+            } else {
+                return false;
+            }
+        });
+        context.logger_filter(std::move(core_filter));
+
         signal_dispatcher->on<io::context::os_signal>([=, &context](int signum, siginfo_t){
             if(signum == SIGHUP) {
                 COCAINE_LOG_INFO(internal_logger, "resetting logging v2 root logger");
                 *root_logger = get_root_logger(context, config);
             }
         });
         context.signal_hub().listen(signal_dispatcher, io_context);
-        init();
     }
 
-    void on_filter_list(std::future<response::children_subscribe> future) {
-        try {
-            auto filter_ids = future.get();
-            COCAINE_LOG_DEBUG(internal_logger,
-                              "received filter list update, count - {}",
-                              std::get<1>(filter_ids).size());
-            for (auto& id_str : std::get<1>(filter_ids)) {
-                logging::filter_t::id_type filter_id = std::stoull(id_str);
-                listen_for_filter(filter_id);
+    template <class T, class F>
+    struct
+    safe_t {
+        safe_t(std::weak_ptr<T> _self, F _f):
+            self(std::move(_self)),
+            f(std::move(_f))
+        {}
+
+        template<class... Args>
+        auto operator()(Args&&... args) -> void {
+            if(auto lock = self.lock()) {
+                f(std::forward<Args>(args)...);
```

So any callback may be silently ignored just because you give a damn how long does your resources live?

---

As discussed - we can not control circular dependency here and callback invocation, so yes, in case of service shutdown those callbacks will be silently ignored - we are not interested in them.